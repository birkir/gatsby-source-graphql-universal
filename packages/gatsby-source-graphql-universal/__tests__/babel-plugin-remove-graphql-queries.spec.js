const babel = require(`@babel/core`)
const plugin = require(`../src/babel-plugin-remove-graphql-queries.js`)

function matchesSnapshot(query) {
  const { code } = babel.transform(query, {
    presets: [`@babel/preset-react`],
    plugins: [plugin],
  })
  expect(code).toMatchSnapshot()
}

it.todo(
  `Works correctly with the kitchen sink`
  // , () => {
  //   matchesSnapshot(`
  //   import * as React from 'react'
  //   import { graphql, useStaticQuery, StaticQuery } from 'gatsby'

  //   export default () => {
  //     const query = graphql\`{site { siteMetadata { title }}}\`
  //     const siteDescription = useStaticQuery(query)

  //     return (
  //       <StaticQuery
  //         query={graphql\`{site { siteMetadata { title }}}\`}
  //         render={data => (
  //           <div>
  //             <h1>{data.site.siteMetadata.title}</h1>
  //             <p>{siteDescription.site.siteMetadata.description}</p>
  //           </div>
  //         )}
  //       />
  //     )
  //   }
  //   `)
  // }
)

it(`Transforms queries in useStaticQuery`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { graphql, useStaticQuery } from 'gatsby'

  export default () => {
    const siteTitle = useStaticQuery(graphql\`{site { siteMetadata { title }}}\`)

    return (
      <h1>{siteTitle.site.siteMetadata.title}</h1>
    )
  }
  `)
})

it(`Transforms exported queries in useStaticQuery`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { graphql, useStaticQuery } from 'gatsby'

  export default () => {
    const data = useStaticQuery(query)

    return (
      <>
        <h1>{data.site.siteMetadata.title}</h1>
        <p>{data.site.siteMetadata.description}</p>
      </>
    )
  }

  export const query = graphql\`{site { siteMetadata { title }}}\`
  `)
})

it(`Transforms queries defined in own variable in useStaticQuery`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { graphql, useStaticQuery } from 'gatsby'

  export default () => {
    const query = graphql\`{site { siteMetadata { title }}}\`
    const siteTitle = useStaticQuery(query)

    return (
      <h1>{siteTitle.site.siteMetadata.title}</h1>
    )
  }
  `)
})

it(`Transforms queries and preserves destructuring in useStaticQuery`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { graphql, useStaticQuery } from 'gatsby'

  export default () => {
    const query = graphql\`{site { siteMetadata { title }}}\`
    const { site } = useStaticQuery(query)

    return (
      <h1>{site.siteMetadata.title}</h1>
    )
  }
  `)
})

it(`Transforms queries and preserves variable type in useStaticQuery`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { graphql, useStaticQuery } from 'gatsby'

  export default () => {
    const query = graphql\`{site { siteMetadata { title }}}\`
    let { site } = useStaticQuery(query)

    return (
      <h1>{site.siteMetadata.title}</h1>
    )
  }
  `)
})

it(`Transformation does not break custom hooks`, () => {
  matchesSnapshot(`
  import React from "react"
  import { graphql, useStaticQuery } from "gatsby"

  const useSiteMetadata = () => {
    const data = useStaticQuery(graphql\`{site { siteMetadata { title }}}\`)
    return data.site.siteMetadata
  }

  export default () => {
    const siteMetadata = useSiteMetadata()

    return <h1>{site.siteMetadata.title}</h1>
  }

  `)
})

it(`Transforms only the call expression in useStaticQuery`, () => {
  matchesSnapshot(`
  import React from "react"
  import { graphql, useStaticQuery } from "gatsby"

  const useSiteMetadata = () => {
    return useStaticQuery(
      graphql\`{site { siteMetadata { title }}}\`
    ).site.siteMetadata
  }

  export default () => {
    const siteMetadata = useSiteMetadata()

    return <h1>{siteMetadata.title}</h1>
  }
  `)
})

it(`Only runs transforms if useStaticQuery is imported from gatsby`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { graphql } from 'gatsby'

  export default () => {
    const query = graphql\`{site { siteMetadata { title }}}\`
    const siteTitle = useStaticQuery(query)

    return (
      <h1>{siteTitle.site.siteMetadata.title}</h1>
    )
  }
  `)
})

it(`Allow alternative import of useStaticQuery`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import * as Gatsby from 'gatsby'

  export default () => {
    const query = Gatsby.graphql\`{site { siteMetadata { title }}}\`
    const siteTitle = Gatsby.useStaticQuery(query)

    return (
      <h1>{siteTitle.site.siteMetadata.title}</h1>
    )
  }
  `)
})

it(`Transforms queries in <StaticQuery>`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { graphql, StaticQuery } from 'gatsby'

  export default () => (
    <StaticQuery
      query={graphql\`{site { siteMetadata { title }}}\`}
      render={data => <div>{data.site.siteMetadata.title}</div>}
    />
  )
  `)
})

it(`Transforms queries defined in own variable in <StaticQuery>`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { graphql, StaticQuery } from 'gatsby'

  const query = graphql\`{site { siteMetadata { title }}}\`

  export default () => (
    <StaticQuery
      query={query}
      render={data => <div>{data.site.siteMetadata.title}</div>}
    />
  )
  `)
})

it(`transforms exported variable queries in <StaticQuery>`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { graphql, StaticQuery } from 'gatsby'

  export const query = graphql\`{site { siteMetadata { title }}}\`

  export default () => (
    <StaticQuery
      query={query}
      render={data => <div>{data.site.siteMetadata.title}</div>}
    />
  )
  `)
})

it(`Transforms queries in page components`, () => {
  matchesSnapshot(`
  import { graphql } from 'gatsby'

  export const query = graphql\`
     {
       site { siteMetadata { title }}
     }
  \`
  `)
})

it(`allows the global tag`, () => {
  matchesSnapshot(
    `
  export const query = graphql\`
     {
       site { siteMetadata { title }}
     }
  \`
  `
  )
})

it(`distinguishes between the right tags`, () => {
  matchesSnapshot(
    `
  const foo = styled('div')\`
     {
       $\{foo}
     }
  \`

  const pulse = keyframes\`
    0% {
      transform: scale(1);
      animation-timing-function: ease-in;
    }
    25% {
      animation-timing-function: ease-out;
      transform: scale(1.05);
    }
    50% {
      transform: scale(1.12);
      animation-timing-function: ease-in;
    }
    to {
      transform: scale(1);
      animation-timing-function: ease-out;
    }
  \`;

  export const query = graphql\`
     {
       site { siteMetadata { title }}
     }
  \`
  `
  )
})

it(`handles import aliasing`, () => {
  matchesSnapshot(
    `
  import { graphql as gql } from 'gatsby'

  export const query = gql\`
     {
       site { siteMetadata { title }}
     }
  \`
  `
  )
})

it(`handles require`, () => {
  matchesSnapshot(
    `
  const { graphql } = require('gatsby')

  export const query = graphql\`
     {
       site { siteMetadata { title }}
     }
  \`
  `
  )
})

it(`handles require namespace`, () => {
  matchesSnapshot(
    `
  const Gatsby = require('gatsby')

  export const query = Gatsby.graphql\`
     {
       site { siteMetadata { title }}
     }
  \`
  `
  )
})
it(`handles require alias`, () => {
  matchesSnapshot(
    `
  const { graphql: gql } = require('gatsby')

  export const query = gql\`
     {
       site { siteMetadata { title }}
     }
  \`
  `
  )
})

it(`Leaves other graphql tags alone`, () => {
  matchesSnapshot(
    `
  import * as React from 'react'
  import { graphql } from 'relay'

  export default () => (
    <div>{data.site.siteMetadata.title}</div>
  )

  export const query = graphql\`
     {
       site { siteMetadata { title }}
     }
  \`
  `
  )
})

it(`Removes all gatsby queries`, () => {
  matchesSnapshot(
    `
  import { graphql } from 'gatsby'

  export default () => (
    <div>{data.site.siteMetadata.title}</div>
  )

  export const siteMetaQuery = graphql\`
    fragment siteMetaQuery on RootQueryType {
      site {
        siteMetadata {
          title
        }
      }
    }
  \`

  export const query = graphql\`
     {
       ...siteMetaQuery
     }
  \`
  `
  )
})

it(`Handles closing StaticQuery tag`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { graphql, StaticQuery } from 'gatsby'

  export default () => (
    <StaticQuery
      query={graphql\`{site { siteMetadata { title }}}\`}
    >
      {data => <div>{data.site.siteMetadata.title}</div>}
    </StaticQuery>
  )
  `)
})

it(`Doesn't add data import for non static queries`, () => {
  matchesSnapshot(`
  import * as React from 'react'
  import { StaticQuery, graphql } from "gatsby"

  const Test = () => (
    <StaticQuery
      query={graphql\`
      {
        site {
          siteMetadata {
            title
          }
        }
      }
      \`}
      render={data => <div>{data.site.siteMetadata.title}</div>}
    />
  )

  export default Test

  export const fragment = graphql\`
    fragment MarkdownNodeFragment on MarkdownRemark {
      html
    }
  \`
  `)
})
