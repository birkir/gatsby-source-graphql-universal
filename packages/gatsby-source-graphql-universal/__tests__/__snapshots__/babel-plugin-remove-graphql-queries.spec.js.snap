// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Allow alternative import of useStaticQuery 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
import * as Gatsby from 'gatsby';
export default (() => {
  const query = {
    id: \\"426988268\\",
    source: \\"{site{siteMetadata{title}}}\\",

    toString() {
      return this.id;
    }

  };
  const siteTitle = staticQueryData.data;
  return /*#__PURE__*/React.createElement(\\"h1\\", null, siteTitle.site.siteMetadata.title);
});"
`;

exports[`Doesn't add data import for non static queries 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
import { StaticQuery } from \\"gatsby\\";

const Test = () => /*#__PURE__*/React.createElement(StaticQuery, {
  query: {
    id: \\"426988268\\",
    source: \\"{site{siteMetadata{title}}}\\",

    toString() {
      return this.id;
    }

  },
  render: data => /*#__PURE__*/React.createElement(\\"div\\", null, data.site.siteMetadata.title),
  data: staticQueryData
});

export default Test;
export const fragment = {
  id: \\"4176178832\\",
  source: \\"fragment MarkdownNodeFragment on MarkdownRemark{html}\\",

  toString() {
    return this.id;
  }

};"
`;

exports[`Handles closing StaticQuery tag 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
import { StaticQuery } from 'gatsby';
export default (() => /*#__PURE__*/React.createElement(StaticQuery, {
  query: {
    id: \\"426988268\\",
    source: \\"{site{siteMetadata{title}}}\\",

    toString() {
      return this.id;
    }

  },
  data: staticQueryData
}, data => /*#__PURE__*/React.createElement(\\"div\\", null, data.site.siteMetadata.title)));"
`;

exports[`Leaves other graphql tags alone 1`] = `
"import * as React from 'react';
import { graphql } from 'relay';
export default (() => /*#__PURE__*/React.createElement(\\"div\\", null, data.site.siteMetadata.title));
export const query = graphql\`
     {
       site { siteMetadata { title }}
     }
  \`;"
`;

exports[`Only runs transforms if useStaticQuery is imported from gatsby 1`] = `
"import * as React from 'react';
export default (() => {
  const query = {
    id: \\"426988268\\",
    source: \\"{site{siteMetadata{title}}}\\",

    toString() {
      return this.id;
    }

  };
  const siteTitle = useStaticQuery(query);
  return /*#__PURE__*/React.createElement(\\"h1\\", null, siteTitle.site.siteMetadata.title);
});"
`;

exports[`Removes all gatsby queries 1`] = `
"export default (() => /*#__PURE__*/React.createElement(\\"div\\", null, data.site.siteMetadata.title));
export const siteMetaQuery = {
  id: \\"504726680\\",
  source: \\"fragment siteMetaQuery on RootQueryType{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};
export const query = {
  id: \\"3211238532\\",
  source: \\"{...siteMetaQuery}\\",

  toString() {
    return this.id;
  }

};"
`;

exports[`Transformation does not break custom hooks 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import React from \\"react\\";

const useSiteMetadata = () => {
  const data = staticQueryData.data;
  return data.site.siteMetadata;
};

export default (() => {
  const siteMetadata = useSiteMetadata();
  return /*#__PURE__*/React.createElement(\\"h1\\", null, site.siteMetadata.title);
});"
`;

exports[`Transforms exported queries in useStaticQuery 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
export default (() => {
  const data = staticQueryData.data;
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(\\"h1\\", null, data.site.siteMetadata.title), /*#__PURE__*/React.createElement(\\"p\\", null, data.site.siteMetadata.description));
});
export const query = {
  id: \\"426988268\\",
  source: \\"{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};"
`;

exports[`Transforms only the call expression in useStaticQuery 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import React from \\"react\\";

const useSiteMetadata = () => {
  return staticQueryData.data.site.siteMetadata;
};

export default (() => {
  const siteMetadata = useSiteMetadata();
  return /*#__PURE__*/React.createElement(\\"h1\\", null, siteMetadata.title);
});"
`;

exports[`Transforms queries and preserves destructuring in useStaticQuery 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
export default (() => {
  const query = {
    id: \\"426988268\\",
    source: \\"{site{siteMetadata{title}}}\\",

    toString() {
      return this.id;
    }

  };
  const {
    site
  } = staticQueryData.data;
  return /*#__PURE__*/React.createElement(\\"h1\\", null, site.siteMetadata.title);
});"
`;

exports[`Transforms queries and preserves variable type in useStaticQuery 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
export default (() => {
  const query = {
    id: \\"426988268\\",
    source: \\"{site{siteMetadata{title}}}\\",

    toString() {
      return this.id;
    }

  };
  let {
    site
  } = staticQueryData.data;
  return /*#__PURE__*/React.createElement(\\"h1\\", null, site.siteMetadata.title);
});"
`;

exports[`Transforms queries defined in own variable in <StaticQuery> 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
import { StaticQuery } from 'gatsby';
const query = {
  id: \\"426988268\\",
  source: \\"{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};
export default (() => /*#__PURE__*/React.createElement(StaticQuery, {
  query: query,
  render: data => /*#__PURE__*/React.createElement(\\"div\\", null, data.site.siteMetadata.title),
  data: staticQueryData
}));"
`;

exports[`Transforms queries defined in own variable in useStaticQuery 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
export default (() => {
  const query = {
    id: \\"426988268\\",
    source: \\"{site{siteMetadata{title}}}\\",

    toString() {
      return this.id;
    }

  };
  const siteTitle = staticQueryData.data;
  return /*#__PURE__*/React.createElement(\\"h1\\", null, siteTitle.site.siteMetadata.title);
});"
`;

exports[`Transforms queries in <StaticQuery> 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
import { StaticQuery } from 'gatsby';
export default (() => /*#__PURE__*/React.createElement(StaticQuery, {
  query: {
    id: \\"426988268\\",
    source: \\"{site{siteMetadata{title}}}\\",

    toString() {
      return this.id;
    }

  },
  render: data => /*#__PURE__*/React.createElement(\\"div\\", null, data.site.siteMetadata.title),
  data: staticQueryData
}));"
`;

exports[`Transforms queries in page components 1`] = `
"export const query = {
  id: \\"426988268\\",
  source: \\"{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};"
`;

exports[`Transforms queries in useStaticQuery 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
export default (() => {
  const siteTitle = staticQueryData.data;
  return /*#__PURE__*/React.createElement(\\"h1\\", null, siteTitle.site.siteMetadata.title);
});"
`;

exports[`allows the global tag 1`] = `
"export const query = {
  id: \\"426988268\\",
  source: \\"{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};"
`;

exports[`distinguishes between the right tags 1`] = `
"const foo = styled('div')\`
     {
       \${foo}
     }
  \`;
const pulse = keyframes\`
    0% {
      transform: scale(1);
      animation-timing-function: ease-in;
    }
    25% {
      animation-timing-function: ease-out;
      transform: scale(1.05);
    }
    50% {
      transform: scale(1.12);
      animation-timing-function: ease-in;
    }
    to {
      transform: scale(1);
      animation-timing-function: ease-out;
    }
  \`;
export const query = {
  id: \\"426988268\\",
  source: \\"{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};"
`;

exports[`handles import aliasing 1`] = `
"export const query = {
  id: \\"426988268\\",
  source: \\"{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};"
`;

exports[`handles require 1`] = `
"export const query = {
  id: \\"426988268\\",
  source: \\"{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};"
`;

exports[`handles require alias 1`] = `
"export const query = {
  id: \\"426988268\\",
  source: \\"{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};"
`;

exports[`handles require namespace 1`] = `
"export const query = {
  id: \\"426988268\\",
  source: \\"{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};"
`;

exports[`transforms exported variable queries in <StaticQuery> 1`] = `
"import staticQueryData from \\"public/static/d/426988268.json\\";
import * as React from 'react';
import { StaticQuery } from 'gatsby';
export const query = {
  id: \\"426988268\\",
  source: \\"{site{siteMetadata{title}}}\\",

  toString() {
    return this.id;
  }

};
export default (() => /*#__PURE__*/React.createElement(StaticQuery, {
  query: query,
  render: data => /*#__PURE__*/React.createElement(\\"div\\", null, data.site.siteMetadata.title),
  data: staticQueryData
}));"
`;
